File: TSP/solver.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <stdbool.h>
#include <time.h>
#include <limits.h>

#define MAX_NODES 1000
#define MAX_RUNS 5
#define FILEPATH "TSP_instances/xqf131.tsp"
#define OPT_FILEPATH "TSP_instances/xqf131.tour" // Optional set to NULL if theres none.
#define NAME "xqf131"

typedef struct {
    long long int x, y; // int may be small for some coordinates
} Point;

typedef struct {
    int* tour;
    int dist;
} TourResult;

typedef struct {
    int node;
    int distance;
} Neighbor;

Point* parse_tsp_file(const char* file_path, int* num_points);
int* parse_tour_file(const char* file_path, int* num_points);
int create_tour_file();
void update_tour_file(const int* tour, int num_nodes, int dist, double time, int tourfile_number);
int** pre_process(Point* points, int num_points);
int calculate_distance(Point p1, Point p2);
int calculate_tour_length(const int* tour, int n, int** distances);
int calculate_tour_distance(Point* points, const int* tour, int n);
TourResult two_opt_swap(int** distances, const int* initial_tour, int n);
int* two_opt_reverse(int** distances, const int* initial_tour, int n);
TourResult two_opt_and_swap(int** distances, const int* initial_tour, int n);
int* nearest_neighbor(int** distances, int n, int initial_point);
void free_distances(int** distances, int num_points);
int compare_neighbors(const void* a, const void* b);

int compare_neighbors(const void* a, const void* b) {
    Neighbor* na = (Neighbor*)a;
    Neighbor* nb = (Neighbor*)b;
    if (na->distance != nb->distance) return na->distance - nb->distance;
    return na->node - nb->node;
}

Point* parse_tsp_file(const char* file_path, int* num_points) {
    FILE* file = fopen(file_path, "r");
    if (!file) {
        printf("Error: Unable to open file %s\n", file_path);
        return NULL;
    }

    Point* points = malloc(MAX_NODES * sizeof(Point));
    if (!points) {
        fclose(file);
        return NULL;
    }

    char line[256];
    bool node_section = false;
    int count = 0;

    while (fgets(line, sizeof(line), file)) {
        if (strstr(line, "NODE_COORD_SECTION")) {
            node_section = true;
            continue;
        }
        if (strstr(line, "EOF") || line[0] == '\0') break;

        if (node_section) {
            int id;
            float x, y;
            if (sscanf(line, "%d %f %f", &id, &x, &y) == 3) {
                if (count >= MAX_NODES) {
                    printf("Error: Too many points\n");
                    free(points);
                    fclose(file);
                    return NULL;
                }
                points[count].x = round(x);
                points[count].y = round(y);
                count++;
            }
        }
    }

    *num_points = count;
    fclose(file);
    return points;
}

int* parse_tour_file(const char* file_path, int* num_points) {
    FILE* file = fopen(file_path, "r");
    if (!file) {
        return NULL;
    }

    int* tour = malloc(MAX_NODES * sizeof(int));
    if (!tour) {
        fclose(file);
        return NULL;
    }

    char line[256];
    bool tour_section = false;
    int count = 0;

    while (fgets(line, sizeof(line), file)) {
        if (strstr(line, "TOUR_SECTION")) {
            tour_section = true;
            continue;
        }
        if (strcmp(line, "-1\n") == 0) break;

        if (tour_section) {
            int point;
            if (sscanf(line, "%d", &point) == 1) {
                if (count >= MAX_NODES) {
                    printf("Error: Too many tour points\n");
                    free(tour);
                    fclose(file);
                    return NULL;
                }
                tour[count++] = point - 1; // 1-based to 0-based
            }
        }
    }

    fclose(file);
    return tour;
}

int create_tour_file() {
    int counter = 1;
    char filepath[256];
    snprintf(filepath, sizeof(filepath), "TSP_results/%s_%d.tour", NAME, counter); 

    while (fopen(filepath, "r")) {
        counter++;
        snprintf(filepath, sizeof(filepath), "TSP_results/%s_%d.tour", NAME, counter);
    }

    FILE* f = fopen(filepath, "w");
    if (!f) {
        printf("Error: Unable to create file %s\n", filepath);
        return 0;
    }

    fclose(f);
    printf("Tour file created in %s\n", filepath);

    return counter;
}

void update_tour_file(const int* tour, int num_nodes, int dist, double time, int tourfile_number) {
    char filepath[256];
    snprintf(filepath, sizeof(filepath), "TSP_results/%s_%d.tour", NAME, tourfile_number);

    FILE* f = fopen(filepath, "w");
    if (!f) {
        printf("Error: Unable to update file %s\n", filepath);
        return;
    }

    fprintf(f, "NAME: %s\n", NAME);
    fprintf(f, "COMMENT: Tour length %d, total time %.2f seconds\n", dist, time);
    fprintf(f, "TYPE: TOUR\n");
    fprintf(f, "DIMENSION: %d\n", num_nodes);
    fprintf(f, "TOUR_SECTION\n");
    for (int i = 0; i < num_nodes; i++) fprintf(f, "%d\n", tour[i] + 1);
    fprintf(f, "-1\nEOF\n");
    fclose(f);
    printf("Tour updated in %s\n", filepath);
}

int calculate_distance(Point p1, Point p2) {
    return round(sqrt((p1.x - p2.x) * (p1.x - p2.x) + (p1.y - p2.y) * (p1.y - p2.y)));
}

int** pre_process(Point* points, int num_points) {
    int** distances = malloc(num_points * sizeof(int*));
    Neighbor* temp = malloc(num_points * sizeof(Neighbor));
    for (int i = 0; i < num_points; i++) {
        distances[i] = malloc(num_points * sizeof(int));
    }

    for (int i = 0; i < num_points; i++) {
        int temp_count = 0;
        for (int j = 0; j < num_points; j++) {
            if (i == j) {
                distances[i][j] = 0;
                continue;
            }
            temp[temp_count].node = j;
            temp[temp_count].distance = calculate_distance(points[i], points[j]);
            temp_count++;
        }
        qsort(temp, temp_count, sizeof(Neighbor), compare_neighbors);
        for (int k = 0; k < temp_count; k++) {
            distances[i][temp[k].node] = temp[k].distance;
        }
    }
    free(temp);
    return distances;
}

int calculate_tour_length(const int* tour, int n, int** distances) {
    int length = 0;
    for (int i = 0; i < n; i++) {
        int current = tour[i];
        int next = tour[(i + 1) % n];
        length += distances[current][next];
    }
    return length;
}

int calculate_tour_distance(Point* points, const int* tour, int n) {
    int total = 0;
    for (int i = 0; i < n - 1; i++) {
        total += calculate_distance(points[tour[i]], points[tour[i + 1]]);
    }
    total += calculate_distance(points[tour[n - 1]], points[tour[0]]);
    return total;
}

TourResult two_opt_swap(int** distances, const int* initial_tour, int n) {
    int* best_tour = malloc(n * sizeof(int));
    memcpy(best_tour, initial_tour, n * sizeof(int));
    int shortest_dist = calculate_tour_length(best_tour, n, distances);
    bool improved = true;

    while (improved) {
        improved = false;
        for (int i = 0; i < n - 1; i++) {
            for (int j = i + 2; j < n; j++) {
                int a = best_tour[i], b = best_tour[(i + 1) % n];
                int c = best_tour[j], d = best_tour[(j + 1) % n];
                int current_dist = distances[a][b] + distances[c][d];
                int new_dist = distances[a][c] + distances[b][d];
                if (new_dist < current_dist) {
                    int delta = new_dist - current_dist;
                    if (delta < 0) {
                        for (int k = 0; k < (j - i) / 2; k++) {
                            int temp = best_tour[i + 1 + k];
                            best_tour[i + 1 + k] = best_tour[j - k];
                            best_tour[j - k] = temp;
                        }
                        shortest_dist += delta;
                        improved = true;
                    }
                }
            }
        }
    }

    TourResult result = {best_tour, shortest_dist};
    return result;
}

int* two_opt_reverse(int** distances, const int* initial_tour, int n) {
    int* best_tour = malloc(n * sizeof(int));
    memcpy(best_tour, initial_tour, n * sizeof(int));
    int longest_dist = calculate_tour_length(best_tour, n, distances);
    bool improved = true;

    while (improved) {
        improved = false;
        for (int i = 0; i < n - 1; i++) {
            for (int j = i + 2; j < n; j++) {
                int* temp_tour = malloc(n * sizeof(int));
                memcpy(temp_tour, best_tour, n * sizeof(int));
                for (int k = 0; k < (j - i) / 2; k++) {
                    int temp = temp_tour[i + 1 + k];
                    temp_tour[i + 1 + k] = temp_tour[j - k];
                    temp_tour[j - k] = temp;
                }
                int new_dist = calculate_tour_length(temp_tour, n, distances);
                if (new_dist > longest_dist) {
                    free(best_tour);
                    best_tour = temp_tour;
                    longest_dist = new_dist;
                    improved = true;
                } else {
                    free(temp_tour);
                }
            }
        }
    }
    return best_tour;
}

TourResult two_opt_and_swap(int** distances, const int* initial_tour, int n) {
    int* worsened_tour = two_opt_reverse(distances, initial_tour, n);
    TourResult best = two_opt_swap(distances, worsened_tour, n);
    free(worsened_tour);
    int* best_tour = best.tour;
    int shortest_dist = best.dist;
    int* current_tour = malloc(n * sizeof(int));
    memcpy(current_tour, best_tour, n * sizeof(int));
    bool improved = true;

    while (improved) {
        improved = false;
        for (int i = 0; i < n - 1; i++) {
            for (int j = 0; j < n; j++) {
                if (i != j) {
                    int temp = current_tour[i];
                    current_tour[i] = current_tour[j];
                    current_tour[j] = temp;
                    TourResult temp_result = two_opt_swap(distances, current_tour, n);
                    if (temp_result.dist < shortest_dist) {
                        free(best_tour);
                        best_tour = temp_result.tour;
                        memcpy(current_tour, best_tour, n * sizeof(int));
                        shortest_dist = temp_result.dist;
                        improved = true;
                        printf("2-opt improvement: %d\n", shortest_dist);
                        break;
                    } else {
                        free(temp_result.tour);
                    }
                }
            }
            if (improved) break;
        }
    }
end:
    free(current_tour);
    TourResult result = {best_tour, shortest_dist};
    return result;
}

int* nearest_neighbor(int** distances, int n, int initial_point) {
    int* tour = malloc(n * sizeof(int));
    bool* visited = calloc(n, sizeof(bool));
    int tour_size = 0;

    tour[tour_size++] = 0;
    visited[0] = true;
    if (initial_point != 0 && initial_point >= 0 && initial_point < n) {
        tour[tour_size++] = initial_point;
        visited[initial_point] = true;
    }
    int last_element = initial_point != 0 ? initial_point : 0;

    while (tour_size < n) {
        int min_dist = INT_MAX;
        int next_node = -1;
        for (int j = 0; j < n; j++) {
            if (!visited[j] && distances[last_element][j] < min_dist) {
                min_dist = distances[last_element][j];
                next_node = j;
            }
        }
        if (next_node != -1) {
            tour[tour_size++] = next_node;
            visited[next_node] = true;
            last_element = next_node;
        }
    }

    printf("[");
    for (int i = 0; i < n; i++) {
        printf("%d", tour[i]);
        if (i < n - 1) printf(", ");
    }
    printf("]\n");

    free(visited);
    return tour;
}

void free_distances(int** distances, int num_points) {
    for (int i = 0; i < num_points; i++) free(distances[i]);
    free(distances);
}

TourResult solve_tsp(Point* points, int num_points) { 
    clock_t start = clock();
    int** distances = pre_process(points, num_points);
    int initial_point = 1;
    int limit = initial_point + MAX_RUNS;

    int* best_tour = NULL;
    int shortest_dist = INT_MAX;

    int tourfile_number = create_tour_file();

    while (initial_point < limit) {
        printf("current run: [%d], time: %.2f seconds\n", initial_point, (double)(clock() - start) / CLOCKS_PER_SEC);

        int* initial_tour = nearest_neighbor(distances, num_points, initial_point);
        TourResult result = two_opt_and_swap(distances, initial_tour, num_points);

        free(initial_tour);
        if (result.dist < shortest_dist) {
            if (best_tour) free(best_tour);
            best_tour = result.tour;
            shortest_dist = result.dist;
            printf("New shortest dist: %d\n", shortest_dist);
            printf("Saving tour...\n");
            update_tour_file(best_tour, num_points, shortest_dist, (double)(clock() - start) / CLOCKS_PER_SEC, tourfile_number);
            printf("Saved\n");
        } else {
            free(result.tour);
            update_tour_file(best_tour, num_points, shortest_dist, (double)(clock() - start) / CLOCKS_PER_SEC, tourfile_number);
        }

        initial_point++;
    }


    printf("Total time: %.2f seconds\n", (double)(clock() - start) / CLOCKS_PER_SEC);

    free_distances(distances, num_points);
    TourResult result = {best_tour, shortest_dist};
    return result;
}

int main() {
    int num_points;
    Point* tsp_points = parse_tsp_file(FILEPATH, &num_points);
    if (!tsp_points) {
        printf("Failed to parse TSP file %s\n", FILEPATH);
        return 1;
    }
    int opt_num_points;
    int* opt_tour = parse_tour_file(OPT_FILEPATH, &opt_num_points);
    int opt_dist = 0;
    if (opt_tour != NULL) {
        opt_dist = calculate_tour_distance(tsp_points, opt_tour, num_points);
    }
    printf("Optimal distance: %d\n", opt_dist);

    TourResult result = solve_tsp(tsp_points, num_points);
    int* tour = result.tour;
    int dist = result.dist;

    printf("Optimal distance: %d\n", opt_dist);
    printf("Best found distance: %d\n", dist);

    free(tour);
    free(opt_tour);
    free(tsp_points);
    return 0;
}

File: TSP/solver.py
import math
import time
import os

#  Global variables
MAX_RUNS = 5
NAME = "pr76"
FILEPATH = "TSP_instances/pr76.tsp"
OPT_FILEPATH = "TSP_instances/pr76.opt.tour" #  Optional set to None if theres none.


def parse_tsp_file(file_path):
    coordinates = []

    with open(file_path, 'r') as file:
        lines = file.readlines()
        file.close()

    node_coord_section = False

    for line in lines:
        line = line.strip()

        if line == "NODE_COORD_SECTION":
            node_coord_section = True
            continue
        if line == "EOF":
            break

        if node_coord_section:
            parts = line.split()
            if len(parts) >= 3:
                x, y = round(float(parts[1])), round(float(parts[2]))
                coordinates.append((x, y))

    return coordinates


def parse_tour_file(file_path):
    if file_path is None:
        return None

    coordinates_points = []

    with open(file_path, 'r') as file:
        lines = file.readlines()
        file.close()

    tour_section = False

    for line in lines:
        line = line.strip()

        if line == "TOUR_SECTION":
            tour_section = True
            continue
        if line == "-1":
            break

        if tour_section:
            parts = line.split()
            if len(parts) == 1:
                point = int(parts[0]) - 1
                coordinates_points.append(point)

    return coordinates_points


def create_tour_file():
    counter = 1

    while True:
        filepath = f"TSP_results/{NAME}_{counter}.tour"
        if not os.path.exists(filepath):
            with open(filepath, "w") as file:
                file.close()
            break
        else:
            counter += 1

    print(f"Tour file created in {filepath}")

    return counter


def update_tour_file(tour, dist, time, tourfile_number):
    filepath = f"TSP_results/{NAME}_{tourfile_number}.tour"

    with open(filepath, "w") as file:
        file.write(f"NAME: {NAME}\n")
        file.write(f"COMMENT: Tour length {dist}, total time {time} seconds\n")
        file.write("TYPE: TOUR\n")
        file.write(f"DIMENSION: {len(tour)}\n")
        file.write("TOUR_SECTION\n")

        for point in tour:
            file.write(f"{point+1}\n")
        file.write("-1\n")
        file.write("EOF\n")
        file.close()

    print(f"Tour updated in {filepath}")


def calculate_distance(p1, p2):
    return round(math.sqrt((p1[0] - p2[0])**2 + (p1[1] - p2[1])**2))


def calculate_tour_length(tour, processed_distances):
    n = len(tour)
    length = 0
    for i in range(n):
        current = tour[i]
        next = tour[(i + 1) % n]
        length += processed_distances[current][next]

    return length


def calculate_tour_distance(points, tour):
    total_distance = 0

    for idx in range(len(tour) - 1):
        point1 = tour[idx]
        point2 = tour[idx + 1]

        initial_point = points[point1]
        target_point = points[point2]

        distance = calculate_distance(initial_point, target_point)
        total_distance += distance

    total_distance += calculate_distance(points[tour[-1]], points[tour[0]])

    return total_distance


def pre_process(points_position):
    processed_distances = {}

    for idx in range(len(points_position)):
        processed_distances[idx] = {}
        distances = {}

        for other_idx in range(len(points_position)):
            if idx != other_idx:
                distance = calculate_distance(points_position[idx], points_position[other_idx])
                distances[other_idx] = round(distance)

        for kv in sorted(distances.items(), key=lambda kv: (kv[1], kv[0])):
            processed_distances[idx][kv[0]] = kv[1]

    return processed_distances


def two_opt_swap(processed_distances, initial_tour):
    best_tour = initial_tour.copy()
    n = len(best_tour)
    shortest_dist = calculate_tour_length(best_tour, processed_distances)
    improved = True

    while improved:
        improved = False
        for i in range(n - 1):
            for j in range(i + 2, n):
                a, b = best_tour[i], best_tour[(i + 1) % n]
                c, d = best_tour[j], best_tour[(j + 1) % n]

                current_dist = processed_distances[a][b] + processed_distances[c][d]
                new_dist = processed_distances[a][c] + processed_distances[b][d]

                if new_dist < current_dist:
                    delta = new_dist - current_dist
                    if delta < 0:
                        best_tour[i + 1:j + 1] = best_tour[j:i:-1]
                        shortest_dist += delta
                        improved = True
                        break
            if improved:
                break

    return best_tour, shortest_dist


def two_opt_reverse(processed_distances, initial_tour):
    best_tour = initial_tour.copy()
    n = len(best_tour)
    longest_dist = calculate_tour_length(initial_tour, processed_distances)
    improved = True

    while improved:
        improved = False

        for i in range(n - 1):
            for j in range(i + 2, n):
                temp_tour = best_tour.copy()
                temp_tour[i+1:j+1] = temp_tour[j:i:-1]
                new_dist = calculate_tour_length(temp_tour, processed_distances)

                if new_dist > longest_dist:
                    best_tour = temp_tour
                    longest_dist = new_dist
                    improved = True
                    break
            if improved:
                break

    return best_tour


def two_opt_and_swap(processed_distances, initial_tour):
    worsened_tour = two_opt_reverse(processed_distances, initial_tour)
    best_tour, shortest_dist = two_opt_swap(processed_distances, worsened_tour)
    current_tour = best_tour.copy()
    n = len(best_tour)
    improved = True

    while improved:
        improved = False

        for i in range(n - 1):
            for j in range(n):
                if i != j:
                    current_tour[i], current_tour[j] = current_tour[j], current_tour[i]
                    new_tour, new_dist = two_opt_swap(processed_distances, current_tour)
                    if new_dist < shortest_dist:
                        best_tour = new_tour
                        current_tour = best_tour.copy()
                        shortest_dist = new_dist
                        improved = True
                        print(f"2-opt improvement: {shortest_dist}")

    return best_tour, shortest_dist


def nearest_neighbor(processed_distances, initial_point):
    current_tour = {0, initial_point}
    last_element = initial_point

    while len(current_tour) < len(processed_distances):
        for point in processed_distances[last_element]:
            if point not in current_tour:
                current_tour.add(point)
                last_element = point
                break

    print(current_tour)
    return list(current_tour)


def solve_tsp(points):
    start = time.time()
    processed_distances = pre_process(points)
    initial_point = 1
    limit = initial_point + MAX_RUNS

    best_tour = None
    shortest_dist = float("inf")

    tourfile_number = create_tour_file()

    while initial_point < limit:
        print(f"current run: [{initial_point}], time: {round(time.time() - start, 2)} seconds")
        initial_point += 1

        initial_tour = nearest_neighbor(processed_distances, initial_point)
        tour, dist = two_opt_and_swap(processed_distances, initial_tour)

        if dist < shortest_dist:
            shortest_dist = dist
            best_tour = tour

            print(f"New shortest dist: {shortest_dist}")
            print("Saving tour...")
            update_tour_file(best_tour, shortest_dist, round(time.time() - start, 2), tourfile_number)
            print("Saved")
        else:
            update_tour_file(best_tour, shortest_dist, round(time.time() - start, 2), tourfile_number)

    print(f"Total time {round(time.time() - start, 2)} seconds")

    return best_tour, shortest_dist


def main():
    points = parse_tsp_file(FILEPATH)
    opt_tour = parse_tour_file(OPT_FILEPATH)
    opt_dist = 0
    if opt_tour is not None:
        opt_dist = calculate_tour_distance(points, opt_tour)
    print(f"Optimal distance: {opt_dist}")

    tour, dist = solve_tsp(points)

    print(f"Optimal distance: {opt_dist}")
    print(f"Best found distance: {dist}")


if __name__ == "__main__":
    main()
